### ISSUE TITLE
Running Blast python tests in parallel lead to random crashes

### ISSUE THREAD

**Issue by moraaar** [o3de/o3de#9295]
**Describe the bug**
We’ve run Blast python tests on each test mode 10 times and the results show that running tests in parallel modes (`EditorSharedTest` and `EditorParallelTest`) results in random tests failing. 

| Executions | EditorSharedTest | EditorParallelTest | EditorBatchedTest | EditorSingleTest |
| ----------- | ----------- |----------- |----------- |----------- |
| 1 | ok | ok | ok | ok |
| 2 | ok | Failed .F...FF. | ok | ok |
| 3 | Failed ......F. | ok | ok | ok |
| 4 | ok | ok | ok | ok |
| 5 | Failed ....F... | Failed ...F.... | ok | ok |
| 6 | ok | ok | ok | ok |
| 7 | ok | Failed ..F..... | ok | ok |
| 8 | Failed .F...... | ok | ok | ok |
| 9 | Failed ....F... | Failed .......F | ok | ok |
| 10 | ok | Failed ....F... | ok | ok |

They all report the same error, which is an immediate editor crash before it gets to execute the test, and it leaves no log available.

````
        E   Test CRASHED, return code 0xc000041d
        E   ---------------
        E   |  Stacktrace |
        E   ---------------
        E   -- No crash log available --
````

**Steps to reproduce**
1. Build ALL targets of O3DE using `AutomatedTesting` project
2. Run Asset Processor, wait until all assets are processed and close Asset Processor.
3. Edit `o3de\AutomatedTesting\Gem\PythonTests\Blast\TestSuite_Main.py` and set all tests to one test mode: EditorSharedTest, EditorParallelTest, EditorBatchedTest, EditorSingleTest.
4. Open a terminal, go to AutomatedTesting folder and run the following command to run python tests: `o3de\AutomatedTesting>..\python\python.cmd -m pytest .\Gem\PythonTests\Blast\TestSuite_Main.py --build-directory=build\windows\bin\profile\`
5. Run the command 10 times and check results.
6. Go to step 3 and run the tests in another test mode.

**Expected behavior**
All tests pass in all tests modes.

**Actual behavior**
Random failures happen when running tests in parallel modes: EditorSharedTest and EditorParallelTest

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Windows
 - Version 10

**Comment by Kadino** [comment]
Definitely a new type of intermittent failure found by the test tools, which appears to be caused by an undiscovered crash in the Editor.

**Comment by AMZN-Dk** [comment]
Current work around for this issue is to run batched tests, there is an issue with that workaround being addressed here: https://github.com/o3de/o3de/issues/9167

**Comment by Kadino** [comment]
It is becoming obvious that @o3de/sig-testing does not have resources available to proactively investigate the set of parallel contention issues across all other SIGs. Adding the Technical Steering Committee, as this should be a discussion of "if/when" we want SIGs investing in these issues across the product.

The upside would be a moderate reduction in execution times of fully-integrated tests (most of which are in Python). Reducing execution time also reduces total infrastructure demand per test. However instead of a direct reduction in pipeline usage, this would likely result in packing more integration tests into the Automated Review pipeline.

Yet this is a nuanced issue, as a decent subset of these parallel-contention issues will not be customer-facing. For example, using two Editors at once is expected to be extremely unlikely. The majority of parallel contention issues which customers encounter will be between the editor and another O3DE tool, which existing tests already cover. Investing the resources to hold O3DE to an artificially high parallel-stability bar would definitely improve the product, but may not currently be a critical goal.

**Comment by lumberyard-employee-dm** [comment]
Determining the areas where parallel runs of applications causes test failures also aren't obvious as to which SIG group should be  responsible for either adding support for parallel execution within the runtime applications or changing the test to not run in parallel.

In this case the Blast python automated test are maintained by the @o3de/sig-simulation group, but the failure in running those test in parallel is most likely an issue with the the design of the Engine not having parallel runs of the same application accounted for. Multiple Editors writing to the same resource may not be considered out of scope of the team that maintains the Editor @o3de/sig-content as that is not a customer workflow as you mentioned.

If a parallel test failure happens in particular SIG ownership of code, I believe they would have to determine whether to fix any issues with running applications in parallel or deciding whether the test itself should run serially

**Comment by Kadino** [comment]
Reassigning this issue back to @o3de/sig-content to determine their preferred course. This is not a bug in the test framework.

**Comment by monroegm** [comment]
Reassigning to SIG-Simulation to see if as the owner of the tests it wants to run them serially instead of in parallel

**Comment by moraaar** [comment]
The blast tests are already running serially as workaround, this ticket is not to fix blast tests, but to investigate why running them in parallel fail randomly. The blast tests are very simple (open level, run game mode, check entities moved, exit game mode) that brings the question if running tests in parallel is actually reliable at all, either from the editor point of view or the test framework point of view.

I wonder if sig simulation is actually the right sig to investigate and answer these questions. I think the that sig implemented the framework or the feature of "tests running in parallel" has a better chance to take a look at these simple tests and coming to a successful conclusion of why they're failing randomly, as sig simulation doesn't necessarily has the knowledge of how "tests running in parallel" work internally.

Maybe this could be a cross-sig effort, so for now adding back sig-content to continue the discussion.

**Comment by hultonha** [comment]
Thanks for the info @moraaar. I would agree this does seem to be a potential problem with the framework as opposed to a specific test. Perhaps a session with myself, @moraaar and @Kadino could be setup to discuss this in more detail and maybe we could all do a pair (or 'mob' 🙂) programming session to try and get to the bottom of the issue without this ticket continuing to move around? Thanks!

**Comment by hultonha** [comment]
I've setup a call for us to have a catch-up about this, we'll report back after.

**Comment by hultonha** [comment]
After discussing this issue with @moraaar and @Kadino, we have come to the conclusion that as of right now we cannot reliably say that multiple instances of the O3DE Editor can run in parallel (regardless of the testing framework itself). That is to say we can't guarantee everything would remain stable if a user was to run two instances of the O3DE Editor at the same time and the effort to support and maintain this is likely not worth the cost.

As the test framework itself is independent of O3DE, we will retain the functionality for it to run applications in parallel, but those applications must first reliably be able to run in parallel on their own. This is a high bar and difficult to achieve but is definitely possible if prioritized highly enough.

In the meantime we are going to make it clear people should _not_ run O3DE tests in parallel as it is highly likely they will run into determinism issues and crashes (see https://github.com/o3de/o3de/issues/12465). 

Being able to run editor tests in parallel is still something we'd like to support though but we decided there is likely a safer approach (e.g. By using containerization or another mechanism to provide better isolation). Myself and @moraaar will follow-up with the @o3de/sig-build to ask about possible options here.

For the moment I'm closing this ticket as 'won't fix' given the context above.
